homeassistant/components/telegram_bot/__init__.py

```diff
@@ -380,31 +388,62 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     if not domain_config:
         return False
 
+    # create config entries from YAML
+    for config_import in domain_config:
+        trusted_networks: list[IPv4Network] = config_import.get(
+            CONF_TRUSTED_NETWORKS, []
+        )
+        trusted_networks_str: list[str] = (
+            [str(trusted_network) for trusted_network in trusted_networks]
+            if trusted_networks
+            else []
+        )
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={CONF_SOURCE: SOURCE_IMPORT},
+                data={
+                    CONF_PLATFORM: config_import.get(CONF_PLATFORM),
+                    CONF_API_KEY: config_import.get(CONF_API_KEY),
+                    CONF_ALLOWED_CHAT_IDS: config_import.get(CONF_ALLOWED_CHAT_IDS),
+                    ATTR_PARSER: config_import.get(ATTR_PARSER),
+                    CONF_PROXY_URL: config_import.get(CONF_PROXY_URL),
+                    CONF_URL: config_import.get(CONF_URL),
+                    CONF_TRUSTED_NETWORKS: trusted_networks_str,
+                },
+            )
+        )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: TelegramBotConfigEntry) -> bool:
+    """Create the Telegram bot from config entry."""
     platforms = await async_get_loaded_integration(hass, DOMAIN).async_get_platforms(
-        {p_config[CONF_PLATFORM] for p_config in domain_config}
+        {entry.data[CONF_PLATFORM]}
     )
 
-    for p_config in domain_config:
-        # Each platform config gets its own bot
-        bot = await hass.async_add_executor_job(initialize_bot, hass, p_config)
-        p_type: str = p_config[CONF_PLATFORM]
+    bot: Bot = await hass.async_add_executor_job(initialize_bot, hass, entry.data)
+    entry.runtime_data = bot
 
-        platform = platforms[p_type]
+    p_type: str = entry.data[CONF_PLATFORM]
+    platform = platforms[p_type]
 
-        _LOGGER.debug("Setting up %s.%s", DOMAIN, p_type)
-        try:
-            receiver_service = await platform.async_setup_platform(hass, bot, p_config)
-            if receiver_service is False:
-                _LOGGER.error("Failed to initialize Telegram bot %s", p_type)
-                return False
-
-        except Exception:
-            _LOGGER.exception("Error setting up platform %s", p_type)
+    _LOGGER.debug("Setting up %s.%s", DOMAIN, p_type)
+    try:
+        receiver_service = await platform.async_setup_platform(hass, bot, entry)
+        if receiver_service is False:
+            _LOGGER.error("Failed to initialize Telegram bot %s", p_type)
             return False
 
-        notify_service = TelegramNotificationService(
-            hass, bot, p_config.get(CONF_ALLOWED_CHAT_IDS), p_config.get(ATTR_PARSER)
-        )
+    except Exception:
+        _LOGGER.exception("Error setting up platform %s", p_type)
```

Comment:
Side note: It's a bit confusing to say platform, since these are not really platforms. It's just modules of the integration.